plugins/api/src/replica_close.cpp

```diff
@@ -367,27 +380,37 @@ namespace
                 }
             }};
 
-            const auto is_write_operation = (O_RDONLY != (l1desc.dataObjInp->openFlags & O_ACCMODE));
+            const auto update_size = !json_input.contains("update_size") || json_input.at("update_size").get<bool>();
+            const auto compute_checksum =
+                json_input.contains("compute_checksum") && json_input.at("compute_checksum").get<bool>();
+
+            // Close the underlying file object.
+            if (const auto ec = close_physical_object(*_comm, l1desc.l3descInx); ec != 0) {
+                log::api::error("Failed to close file object [error_code={}].", ec);
+                if (is_write_operation && update_status) {
+                    update_replica_status_on_error(*_comm, l1desc);
+                }
+                return ec;
+            }
```

Comment:
Say this succeeds, but then one of the catalog updates that follow fails.

If `replica_close` is invoked again to perform the catalog updates, will `close_physical_object` be a hard stop due to the file on disk being closed?

_I feel I may have asked this question before?_

Reply:
I think it depends on what part of the process failed (e.g. Is the L3 descriptor still valid?) and the underlying storage technology. I haven't tried it yet because I would have to get my operation failure testing tool set up again. Do you want me to try and find out?

In any case, as we discussed, the user has explicitly chosen not to update the replica status and is knowingly attempting an advanced maneuver. Should this occur, I feel like all we can do here is bail with an error and, if applicable, try to unlock the data object.

Reply:
Oh, sorry, I misread the first sentence.

In that case, the L3 descriptor would no longer refer to the same replica (or at least the same open call), so it would probably result in an error. The caller would be responsible for finalizing the data object at that point, in my mind, and they probably should not call `replica_close` again. It would be the same effect as calling `replica_close` and requesting that nothing in the catalog gets updated, right?

In the same way, calling `rsDataObjClose` on failure is (and as near as I can tell always has been) inadvisable because it tries really hard to make sure everything is cleaned up, even in the face of failure (L1 descriptor, locked objects, etc.).

Reply:
> I think it depends on what part of the process failed (e.g. Is the L3 descriptor still valid?) and the underlying storage technology.

I agree. However, I think we have to determine if the L3 descriptor is open before calling `close_physical_object`. We can do that by setting it to 0 or -1 following success. If `replica_close` is called again due to a previous failure, `close_physical_object` will be skipped and the catalog updates will be allowed to execute.

> Do you want me to try and find out?

Yes please. Otherwise, I feel users only way out of this case is `iadmin modrepl`.

Thoughts?

Reply:
So... you're saying if `close_physical_object` succeeds, we should set `l1desc.l3descInx` to -1 or something so that we know that this L1 descriptor is valid but the object has already been closed?

Reply:
Correct.

> In that case, the L3 descriptor would no longer refer to the same replica (or at least the same open call), so it would probably result in an error. The caller would be responsible for finalizing the data object at that point, in my mind, and they probably should not call replica_close again. It would be the same effect as calling replica_close and requesting that nothing in the catalog gets updated, right?

I keep forgetting that we have that API. Seems there are two approaches.

I feel checking the L3 descriptor is very simple/approachable for devs using `replica_close`, so my preference is towards that as the solution, but I'm open to your idea as well. We need to think about the pros/cons for these solutions.

Reply:
Sounds good.

Reply:
I think this has been addressed. To avoid getting too deep into changing `replica_close`, I am going to mark this resolved. As always, we can continue the discussion in a new thread if it seems appropriate. Thanks!